### PR TITLE
gitlab modules: added instance variable

### PR DESCRIPTION
**Pull request by benibr**
##### SUMMARY

This add a new module to use the Gitlab Instance Variables API to set/edit instance wide variables (https://docs.gitlab.com/ee/api/instance_level_ci_variables.html).
This module is based on the code of the already existing `gitlab_group_variable` module (https://github.com/ansible-collections/community.general/blob/main/plugins/modules/gitlab_group_variable.py).

##### ISSUE TYPE

- New Module/Plugin Pull Request

##### COMPONENT NAME

gitlab_instance_variable

### CODE REVIEW

File: plugins/modules/gitlab_instance_variable.py [hunk 1]
```diff
@@ -0,0 +1,393 @@
...
+# Copyright (c) 2020, Florent Madiot (scodeman@scode.io)
+# Copyright (c) 2019, Markus Bergholz (markuman@gmail.com)
+# GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or https://www.gnu.org/licenses/gpl-3.0.txt)
+# SPDX-License-Identifier: GPL-3.0-or-later
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+DOCUMENTATION = r'''
+module: gitlab_instance_variable
+short_description: Creates, updates, or deletes GitLab instance variables
```
Comment (felixfontein) [inline plugins/modules/gitlab_instance_variable.py:15]:
```suggestion
short_description: Creates, updates, or deletes GitLab instance variables
version_added: 7.1.0
```

File: plugins/modules/gitlab_instance_variable.py [hunk 2]
```diff
@@ -0,0 +1,393 @@
...
+  community.general.gitlab_instance_variable:
+    api_url: https://gitlab.com
+    api_token: secret_access_token
+    state: absent
+    vars:
+      ACCESS_KEY_ID: acab1221cbad
+'''
+
+RETURN = r'''
+group_variable:
```
Comment (felixfontein) [inline plugins/modules/gitlab_instance_variable.py:123]: Is `group_variable` the right name for the return values? I thought this module is about instance variables, not group variables.

File: plugins/modules/gitlab_instance_variable.py [hunk 3]
```diff
@@ -0,0 +1,393 @@
...
+    description:
+      - When set to C(true), delete all variables which are not untouched in the task.
+    default: false
+    type: bool
+  vars:
+    description:
+      - When the list element is a simple key-value pair, set masked and protected to false.
+      - When the list element is a dict with the keys I(value), I(masked) and I(protected), the user can
+        have full control about whether a value should be masked, protected or both.
+      - Support for group variables requires GitLab >= 13.0.
```
Comment (felixfontein) [inline plugins/modules/gitlab_instance_variable.py:54]: Group or instance variables?

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution! Please check out the section in CONTRIBUTING.md on adding new modules to the collection. I've added some first comments below.

File: plugins/modules/gitlab_instance_variable.py (hunk 1 above)
Comment (benibr) [reply plugins/modules/gitlab_instance_variable.py:15]: added

File: plugins/modules/gitlab_instance_variable.py [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+options:
+  state:
+    description:
+      - Create or delete instance variable.
+    default: present
+    type: str
+    choices: ["present", "absent"]
+  purge:
+    description:
+      - When set to C(true), delete all variables which are not untouched in the task.
```
Comment (felixfontein) [inline plugins/modules/gitlab_instance_variable.py:48]:
```suggestion
      - When set to C(true), delete all variables which are not mentioned in the task.
```

File: plugins/modules/gitlab_instance_variable.py [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+DOCUMENTATION = r'''
+module: gitlab_instance_variable
+short_description: Creates, updates, or deletes GitLab instance variables
+version_added: 7.1.0
+description:
+  - Creates a instance variable if it does not exist.
+  - When a instance variable does exist, its value will be updated if the values are different.
+  - Support for instance variables requires GitLab >= 13.0.
+  - Variables which are untouched in the playbook, but are not untouched in the GitLab instance,
+    they stay untouched (I(purge) is C(false)) or will be deleted (I(purge) is C(true)).
```
Comment (felixfontein) [inline plugins/modules/gitlab_instance_variable.py:22]:
```suggestion
  - Variables which are not mentioned in the module options, but are present on the GitLab instance,
    will either stay (I(purge=false)) or will be deleted (I(purge=true)).
```

File: plugins/modules/gitlab_instance_variable.py [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+      - Create or delete instance variable.
+    default: present
+    type: str
+    choices: ["present", "absent"]
+  purge:
+    description:
+      - When set to C(true), delete all variables which are not untouched in the task.
+    default: false
+    type: bool
+  vars:
```
Comment (felixfontein) [inline plugins/modules/gitlab_instance_variable.py:51]: In the other modules, `vars` was mostly there for backwards compatibility. I would suggest to remove it in this new module, so you only have to handle `variables`. That will make the module's code simpler.

File: plugins/modules/gitlab_instance_variable.py [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+        protected: true
+        variable_type: env_var
+
+- name: Delete one variable
+  community.general.gitlab_instance_variable:
+    api_url: https://gitlab.com
+    api_token: secret_access_token
+    state: absent
+    vars:
+      ACCESS_KEY_ID: acab1221cbad
```
Comment (felixfontein) [inline plugins/modules/gitlab_instance_variable.py:120]: Also the need to specify a value with `vars` even if `state=absent` is strange :)

**Review by felixfontein — COMMENTED** [review]
Please remember to add tests. Thanks.

File: plugins/modules/gitlab_instance_variable.py (hunk 6 above)
Comment (benibr) [reply plugins/modules/gitlab_instance_variable.py]: thx didn't know that, done

File: plugins/modules/gitlab_instance_variable.py (hunk 5 above)
Comment (benibr) [reply plugins/modules/gitlab_instance_variable.py]: done

File: plugins/modules/gitlab_instance_variable.py (hunk 4 above)
Comment (benibr) [reply plugins/modules/gitlab_instance_variable.py]: done

**Comment by benibr** [comment]
> Please remember to add tests. Thanks.

I had time to add a integration test tasks file which I tested the plugin. Found a bug already with that :-)

**Review by felixfontein — APPROVED** [review]
Looks good to me! I'll merge this for 7.1.0 if nobody objects.

**Comment by felixfontein** [comment]
@benibr thanks for your contribution!

**Pull request merged**